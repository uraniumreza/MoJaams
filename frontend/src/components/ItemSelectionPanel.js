import React, { useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlusCircle, faMinusCircle } from '@fortawesome/free-solid-svg-icons';

const ItemSelectionPanel = ({
  step,
  items,
  request,
  variants,
  goBack,
  addToCart,
  setVariantsForSpecificItem,
}) => {
  const [selectedPizza, setSelectedPizza] = useState();
  const [selectedSize, setSelectedSize] = useState();
  const [selectedQuantity, setSelectedQuantity] = useState(1);

  const initializeStates = () => {
    setSelectedPizza(undefined);
    setSelectedSize(undefined);
    setSelectedQuantity(1);
  };

  const addNewItemToCart = () => {
    const newItem = {
      itemName: selectedPizza,
      itemVariant: selectedSize,
      quantity: selectedQuantity,
    };

    addToCart(newItem);
  };

  return (
    <div className={`full-page ${step === 2 ? 'visible' : ''}`}>
      {step === 2 && (
        <>
          {request.error && 'Error!'}
          {request.loading && 'Loading...'}
          {items.length && (
            <>
              <h4 className="title">Grab your Pizza!</h4>
              <div className="pizza-container">
                {items.map((item, index) => (
                  <div
                    className={`pizza-item ${
                      selectedPizza === item ? 'selected' : ''
                    }`}
                    onClick={() => {
                      setSelectedPizza(item);
                      setVariantsForSpecificItem(item);
                    }}
                    key={index.toString()}
                  >
                    <span>{item}</span>
                  </div>
                ))}
              </div>
            </>
          )}

          {variants.length && (
            <div className={`section ${selectedPizza ? 'visible' : ''}`}>
              <h4 className="title">What size you prefer?</h4>
              <div className="pizza-container">
                {variants.map((variant, index) => (
                  <div
                    className={`pizza-item ${
                      selectedSize === variant ? 'selected' : ''
                    }`}
                    onClick={() => setSelectedSize(variant)}
                    key={index.toString()}
                  >
                    <span>{variant}</span>
                  </div>
                ))}
              </div>
            </div>
          )}

          <div className={`section ${selectedSize ? 'visible' : ''}`}>
            <h4 className="title">How many of them?</h4>
            <div className="counter-container">
              <button
                className="counter-btn"
                title="I want less!"
                onClick={() => {
                  selectedQuantity > 1 &&
                    setSelectedQuantity(selectedQuantity - 1);
                }}
              >
                <FontAwesomeIcon icon={faMinusCircle} />
              </button>
              <span className="counter">{selectedQuantity}</span>
              <button
                className="counter-btn"
                title="I want more!"
                onClick={() => setSelectedQuantity(selectedQuantity + 1)}
              >
                <FontAwesomeIcon icon={faPlusCircle} />
              </button>
            </div>
          </div>

          <div className="navigation-container">
            <button
              className="nav-btn back"
              onClick={() => {
                goBack();
                initializeStates();
              }}
            >
              BACK
            </button>

            {selectedPizza && selectedSize && (
              <button className="nav-btn next" onClick={addNewItemToCart}>
                NEXT
              </button>
            )}
          </div>
        </>
      )}
    </div>
  );
};

export default ItemSelectionPanel;
