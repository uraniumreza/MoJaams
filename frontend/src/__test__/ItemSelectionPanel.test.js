import React from 'react';
import renderer from 'react-test-renderer';

import ItemSelectionPanel from '../components/ItemSelectionPanel';

describe('ItemSelectionPanel - SnapShot test', () => {
  it('renders correctly when there are no items and variants', () => {
    const tree = renderer
      .create(
        <ItemSelectionPanel
          step={2}
          items={[]}
          variants={[]}
          request={{}}
          addToCart={() => {}}
          goBack={() => {}}
          setVariantsForSpecificItem={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when there are 1 item and 2 variants', () => {
    const tree = renderer
      .create(
        <ItemSelectionPanel
          step={2}
          items={['Marinara']}
          variants={['Large', 'Medium']}
          request={{}}
          addToCart={() => {}}
          goBack={() => {}}
          setVariantsForSpecificItem={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when there is a selected item for update', () => {
    const tree = renderer
      .create(
        <ItemSelectionPanel
          step={2}
          items={['Marinara']}
          variants={['Large', 'Medium']}
          request={{}}
          addToCart={() => {}}
          goBack={() => {}}
          setVariantsForSpecificItem={() => {}}
          selectedCartItem={{
            itemName: 'Marinara',
            itemVariant: 'Medium',
            quantity: 2,
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
