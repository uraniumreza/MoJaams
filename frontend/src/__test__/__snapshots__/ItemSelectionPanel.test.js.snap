// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemSelectionPanel - SnapShot test renders correctly when there are 1 item and 2 variants 1`] = `
<div
  className="full-page visible"
>
  <h4
    className="title"
  >
    Grab your Pizza!
  </h4>
  <div
    className="pizza-container"
  >
    <div
      className="pizza-item "
      onClick={[Function]}
    >
      <span>
        Marinara
      </span>
    </div>
  </div>
  <div
    className="section "
  >
    <h4
      className="title"
    >
      What size you prefer?
    </h4>
    <div
      className="pizza-container"
    >
      <div
        className="pizza-item "
        onClick={[Function]}
      >
        <span>
          Large
        </span>
      </div>
      <div
        className="pizza-item "
        onClick={[Function]}
      >
        <span>
          Medium
        </span>
      </div>
    </div>
  </div>
  <div
    className="section "
  >
    <h4
      className="title"
    >
      How many of them?
    </h4>
    <div
      className="counter-container"
    >
      <button
        className="counter-btn"
        onClick={[Function]}
        title="I want less!"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-minus-circle fa-w-16 "
          data-icon="minus-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <span
        className="counter"
      >
        1
      </span>
      <button
        className="counter-btn"
        onClick={[Function]}
        title="I want more!"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-plus-circle fa-w-16 "
          data-icon="plus-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="navigation-container"
  >
    <button
      className="nav-btn back"
      onClick={[Function]}
    >
      BACK
    </button>
  </div>
</div>
`;

exports[`ItemSelectionPanel - SnapShot test renders correctly when there are no items and variants 1`] = `
<div
  className="full-page visible"
>
  0
  0
  <div
    className="section "
  >
    <h4
      className="title"
    >
      How many of them?
    </h4>
    <div
      className="counter-container"
    >
      <button
        className="counter-btn"
        onClick={[Function]}
        title="I want less!"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-minus-circle fa-w-16 "
          data-icon="minus-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <span
        className="counter"
      >
        1
      </span>
      <button
        className="counter-btn"
        onClick={[Function]}
        title="I want more!"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-plus-circle fa-w-16 "
          data-icon="plus-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="navigation-container"
  >
    <button
      className="nav-btn back"
      onClick={[Function]}
    >
      BACK
    </button>
  </div>
</div>
`;

exports[`ItemSelectionPanel - SnapShot test renders correctly when there is a selected item for update 1`] = `
<div
  className="full-page visible"
>
  <h4
    className="title"
  >
    Grab your Pizza!
  </h4>
  <div
    className="pizza-container"
  >
    <div
      className="pizza-item "
      onClick={[Function]}
    >
      <span>
        Marinara
      </span>
    </div>
  </div>
  <div
    className="section "
  >
    <h4
      className="title"
    >
      What size you prefer?
    </h4>
    <div
      className="pizza-container"
    >
      <div
        className="pizza-item "
        onClick={[Function]}
      >
        <span>
          Large
        </span>
      </div>
      <div
        className="pizza-item "
        onClick={[Function]}
      >
        <span>
          Medium
        </span>
      </div>
    </div>
  </div>
  <div
    className="section "
  >
    <h4
      className="title"
    >
      How many of them?
    </h4>
    <div
      className="counter-container"
    >
      <button
        className="counter-btn"
        onClick={[Function]}
        title="I want less!"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-minus-circle fa-w-16 "
          data-icon="minus-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
      <span
        className="counter"
      >
        1
      </span>
      <button
        className="counter-btn"
        onClick={[Function]}
        title="I want more!"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-plus-circle fa-w-16 "
          data-icon="plus-circle"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="navigation-container"
  >
    <button
      className="nav-btn back"
      onClick={[Function]}
    >
      BACK
    </button>
  </div>
</div>
`;
